docs/source/advanced-decision-tree.rst

```diff
@@ -0,0 +1,168 @@
+.. _Decision tree:
+
+=============
+Decision tree
+=============
+
+While this documentation calls the structure traversed by the runner a
+"decision tree," it's more precisely described as a directed graphs. The name
+reflects the most common use case in the bundled tests, not the most complex
+structure.
+
+Tree node fields
+================
+
+A decision tree node has two pointers, to a ``child`` and to a ``next_sibling``.
+On initialisation nodes set them to ``None``.
+
+Child nodes
+-----------
+
+When a node matches received message and processes it without errors,
+runner continues execution by switching to the child.
+
+If ``child`` points to ``None``, runner closes open connections and ends
+execution.
+
+To create loops the ``child`` can point to nodes higher in the tree, including
+to itself.
+You need to use this mechanism to allow receiving arbitrary number of messages.
+
+Sibling nodes
+--------------
+
+The runner uses nodes pointed to by ``next_sibling`` when received message
+doesn't match the current node. When sending messages, runner looks
+into ``next_sibling`` when connection got closed.
+
+You can use this mechanism to either break out of loops or to define
+alternatives in execution.
+
+Advanced decision tree structures
+=================================
+
+As mentioned before, the decision tree allows for non-linear relationship
+between nodes.
+
+Loops
+-----
+
+Test case runner in tlsfuzzer can accept arbitrary number of messages if
+the node points to itself as its child.
+
+For example, to accept zero or more NewSessionTicket messages in
+:term:`TLS` 1.3 connection, the script needs to include the following code:
+
+.. code:: python
+
+    cycle = ExpectNewSessionTicket()
+    node = node.add_child(cycle)
+    node.add_child(cycle)
+
+Servers that send the NewSessionTicket after Finished and before
+any other messages, require the preceding code after
+:py:class:`~tlsfuzzer.expect.ExpectFinished`.
+That handles OpenSSL-using servers and others that behave similarly.
+
+.. note::
+    :term:`TLS` standard does allow sending NewSessionTicket messages at
+    arbitrary times after Finished.
+
+If you want the runner to finish the loop on receipt of ApplicationData
+message write:
```

Comment:
'write' sounds odd, what about: "in the following example the runner finish the loop once ApplicationData message is received."